mitiq/cdr/clifford_training_data.py

```diff
@@ -0,0 +1,514 @@
+                         circuit.
+    """
+    defult_sigma_select = 0.5
+    defult_sigma_replace = 0.5
+    N = int(fraction_non_clifford * total_non_cliff)
+    empty_circuit_copy = empty_circuit.copy()
+    rz_non_cliff_copy = rz_non_cliff.copy()
+    additional_options = additional_options.get('additional_options')
+    if additional_options:
+        if 'sigma_select' in additional_options:
+            sigma_select = additional_options.get('sigma_select')
+            if 'sigma_replace' in additional_options:
+                sigma_replace = additional_options.get('sigma_replace')
+            else:
+                sigma_replace = defult_sigma_replace
+        elif 'sigma_select' not in additional_options:
+            sigma_select = defult_sigma_select
+            if 'sigma_replace' in additional_options:
+                sigma_replace = additional_options.get('sigma_replace')
+            else:
+                raise Exception('additional options must be dicitonary with \
+                                keys containing one or both of '
+                                '"sigma_select" and "sigma_replace" both \
+                                equal to some positive float')
+    else:
+        sigma_select = defult_sigma_select
+        sigma_replace = defult_sigma_replace
+    # Choose non Clifford gates to change according to selection methods:
+    if method_select == 'random':
+        columns_to_change = sample(
+            np.arange(0, total_non_cliff, 1).tolist(), total_non_cliff-N)
+
+    elif method_select == 'probabilistic':
+        non_cliff_angles = rz_non_cliff_copy[2]
+        # form a probability distribution:
+        probabilities = _angle_to_probabilities(non_cliff_angles, sigma_select)
+        prob_choose_gate = [k / sum(probabilities) for k in probabilities]
+        columns_to_change = np.random.choice(
+            np.arange(0, total_non_cliff, 1).tolist(),
+            total_non_cliff - N, replace=False, p=prob_choose_gate)
+
+    else:
+        raise Exception('method_select must = "random", "probabilistic"')
+    rz_non_cliff_selected = rz_non_cliff_copy[:, columns_to_change]
+    # Remove these columns from the circuit data (as they are to be changed
+    # below):
+    rz_non_cliff_copy = np.delete(rz_non_cliff_copy, columns_to_change, axis=1)
+    # Now the non Clifford gates have been selected, we need to decide which
+    # Clifford gate to replace them with.
+    # to store original angles replaced:
+    angles_original = rz_non_cliff_selected[2, :].copy()
+    if method_replace == 'closest':
+        rz_non_cliff_selected[2, :] = _closest_clifford(
+            rz_non_cliff_selected[2, :])
+
+    elif method_replace == 'random':
+        rz_non_cliff_selected[2, :] = _random_clifford(
+            rz_non_cliff_selected[2, :])
+
+    elif method_replace == 'probabilistic':
+        rz_non_cliff_selected[2, :] = _probabilistic_angle_to_clifford(
+            rz_non_cliff_selected[2, :], sigma_replace)
```

Comment:
Seeing this makes me think it's much better to have `method_replace` be a function, not a string. The string is just a placeholder for the function.

Reply:
Can you expand on what you mean by this?

Reply:
Sorry, instead of `method_replace` being a string, it could be the function itself. That way this entire block is just

```python
rz_non_cliff_selected[2, :] = method_replace(...)
```

I think we've discussed something like this before, but this is a minor suggestion for now.